### ISSUE TITLE
Adapter for argmin framework

### ISSUE THREAD

**Issue by relf** [relf/egobox#50]
The `Egor` optimizer should benefit from design developed in [argmin](https://github.com/argmin-rs/argmin) optimiser framework (clean API, observers).

See how we can adapt Egor API or at least ease the switch from `argmin` to `Egor` and conversely

**Comment by stefan-k** [comment]
Hi,

I just came across this issue by accident and found it very interesting. One of my main intents for argmin was to provide a framework which simplifies implementing solvers; however, so far I haven't seen much use of that feature outside of argmin itself. 
Making your solvers compatible with argmin could be interesting for you and your users (because of features such as observers and checkpointing) and for me, because your experience may help identify pain points and shortcomings in argmins design. 
I am not familiar with the design of your optimizer, but if you can manage to implement the `Solver` trait, it should in theory work with argmin's `Executor`. 
With this comment I just wanted to let you know that if you do decide to work on this, I am happy to help out and improve argmin's design to more suit your needs if necessary.

**Comment by relf** [comment]
Hi. Yes, I opened this issue as a reminder for future work on checkpointing and observers. I have not dived into argmin, but definitely hope to do so. 
On another point, internally `egobox` uses wrapped `COBYLA` and/or `SLSQP` (from [nlopt](https://github.com/adwhit/rust-nlopt), I even tried my take on [cobyla](https://github.com/relf/cobyla/) wrapping). A pure-rust approach would be great and I also hope to try `argmin` in that area.

**Comment by stefan-k** [comment]
I wrote a very lengthy reply about why the bound is necessary for checkpointing until I realized that it is actually just a remnant of earlier attempts to implement checkpointing. :facepalm: This is a bug, it works perfectly fine without it and I'll remove it right away. Thanks a lot! :)
Note however that if your solver is not serializable/deserializable, any use of the `Executor`s `checkpointing` method will not compile.
